contributors/design-proposals/kubelet-authorizer.md

```diff
@@ -0,0 +1,119 @@
+  // and may be empty if a specific node cannot be determined.
+  // isNode is true if the user.Info represents an identity issued to a node.
+  IdentifyNode(user.Info) (nodeName string, isNode bool)
+}
+```
+
+The default `NodeIdentifier` implementation:
+* `isNode` - true if the user groups contain the `system:nodes` group
+* `nodeName` - populated if `isNode` is true, and the user name is in the format `system:node:<nodeName>`
+
+## Node authorizer
+
+A new node authorizer will be inserted into the authorization chain:
+* API server authorizer (authorizes "loopback" API clients used by components within the API server)
+* Node authorizer (new)
+* User-configured authorizers... (e.g. ABAC, RBAC, Webhook)
+
+The node authorizer is responsible for ensuring:
+* The request is made by a node (`IdentifyNode()` returns isNode=true)
+* The request is allowed by the rules contained in the default `system:node` cluster role
+
+Requests that do not meet those conditions are forbidden by this authorizer.
+Subsequent authorizers in the chain can run and choose to allow the request.
+
+Requests from identifiable nodes (`IdentifyNode()` returns nodeName != "") for secrets and configmaps are further restricted:
+* Requests for secrets are limited to `get`, and the requested secret must be related to the requesting node by one of the following relationships:
+  * node -> pod -> secret
+  * node -> pod -> pvc -> pv -> secret
+  * node -> pod -> pvc -> storageclass -> secret
+* Requests for configmaps are limited to `get`, and the requested configmap must be related to the requesting node by one of the following relationships:
+  * node -> pod -> configmap
+
+## Node admission
+
+A new node admission plugin is made available that enforces the following:
+
+Limits `create` of node resources by identifiable nodes:
+* only allow the node object corresponding to the node making the API request
+
+Limits `create` of pod resources by identifiable nodes:
+* only allow pods with mirror pod annotations
+* only allow pods with nodeName set to the node making the API request
+* only allow pods with no secret references
+
+Limits `update`,`patch`,`delete` of node and nodes/status resources by identifiable nodes:
+* only allow modifying the node object corresponding to the node making the API request
+
+Limits `update`,`patch`,`delete` of pod and pod/status resources by identifiable nodes:
+* only allow modifying pods with nodeName set to the node making the API request (requires fetching the pod on delete)
+* do not allow removing a mirror pod annotation
+
+## API Changes
+
+None
+
+## RBAC Changes
+
+Currently, the `system:node` cluster role is automatically bound to the `system:nodes` group.
+
+Because the node authorizer accomplishes the same purpose, with the benefit of additional restrictions
+on secret and configmap access, this binding is no longer needed, and will no longer be set up automatically.
```

Comment:
yes... clearly I'm not explaining this well :)

This is how the node authorizer would work, in order:
1. If a request is not from a node, reject
2. If a request is not allowed by the `system:node` rules, reject
3. If a specific node cannot be identified, allow (this is what lets nodes that don't use node-specific identities continue to work with the broad authorization rules in step 2)
4. If resources we care about subdividing access to (secrets, configmaps, etc), determine if the requested object is related to the requesting node, and reject the request if it isn't
5. For other resources, allow

FWIW, this is my current authorizer implementation:

```go
func (r *NodeAuthorizer) Authorize(attrs authorizer.Attributes) (bool, string, error) {
	nodeName, isNode := r.identifier.NodeIdentity(attrs.GetUser())
	if !isNode {
		// reject requests from non-nodes
		return false, "", nil
	}
	if len(nodeName) == 0 && r.requireNodeName {
		// reject requests from unidentifiable nodes if in strict mode
		return false, "unknown node", nil
	}

	if !rbac.RulesAllow(attrs, r.nodeRules...) {
		// reject requests that don't fit within the broad authorization rules
		return false, "", nil
	}

	if len(nodeName) == 0 {
		// allow requests from unidentifiable nodes if not in strict mode
		return true, "", nil
	}

	// subdivide access to specific resources

	if attrs.GetAPIGroup() != api.GroupName {
		// the resources we care about subdividing happen to be in the core API group.
		// access to resources in other API groups is not subdivided.
		return true, "", nil
	}

	switch attrs.GetResource() {
	case "configmaps":
		if attrs.GetVerb() != "get" || len(attrs.GetName()) == 0 || len(attrs.GetNamespace()) == 0 {
			glog.V(2).Infof("NODE DENY: %#v", attrs)
			return false, "can only get individual configmaps", nil
		}
		return r.hasPathFrom(nodeName, ConfigMapNode, attrs.GetNamespace(), attrs.GetName()), "", nil

	case "secrets":
		if attrs.GetVerb() != "get" || len(attrs.GetName()) == 0 || len(attrs.GetNamespace()) == 0 {
			glog.V(2).Infof("NODE DENY: %#v", attrs)
			return false, "can only get individual secrets", nil
		}
		return r.hasPathFrom(nodeName, SecretNode, attrs.GetNamespace(), attrs.GetName()), "", nil

	case "persistentvolumeclaims":
		if attrs.GetVerb() != "get" || len(attrs.GetName()) == 0 || len(attrs.GetNamespace()) == 0 {
			glog.V(2).Infof("NODE DENY: %#v", attrs)
			return false, "can only get individual persistentvolumeclaims", nil
		}
		return r.hasPathFrom(nodeName, PVCNode, attrs.GetNamespace(), attrs.GetName()), "", nil

	case "persistentvolumes":
		if attrs.GetVerb() != "get" || len(attrs.GetName()) == 0 {
			glog.V(2).Infof("NODE DENY: %#v", attrs)
			return false, "can only get individual persistentvolumes", nil
		}
		return r.hasPathFrom(nodeName, PVNode, "", attrs.GetName()), "", nil

	default:
		// Access to other resources is not subdivided
		return true, "", nil
	}
}
```